File: apps/portal-server/src/utils/turbovnc.ts
import { getClusterConfigs } from "@scow/config/build/cluster";
import { getPortalConfig } from "@scow/config/build/portal";
import { join } from "path";

export function getTurboVNCPath(cluster: string) {
  const commonTurboVNCPath = getPortalConfig().turboVNCPath;

  const clusterTurboVNCPath = getClusterConfigs(undefined, undefined, ["hpc"])[cluster].turboVNCPath;

  return clusterTurboVNCPath || commonTurboVNCPath;
}

export function getTurboVNCBinPath(cluster: string, cmd: string) {
  const turboVNCPath = getTurboVNCPath(cluster);

  return join(turboVNCPath, "bin", cmd);
}

const DISPLAY_ID_PORT_DELTA = 5900;

export function parseListOutput(output: string): number[] {
  const ids = [] as number[];
  for (const line of output.split("\n")) {
    if (line.startsWith(":")) {
      const parts = line.split(" ");
      ids.push(parseInt(parts[0].substring(1)));
    }
  }

  return ids;
}

export function parseOtp(stderr: string): string {
  const indicator = "Full control one-time password: ";
  for (const line of stderr.split("\n")) {
    if (line.startsWith(indicator)) {
      return line.substring(indicator.length).trim();
    }
  }

  throw new Error("Error parsing OTP");
}

export function parseDisplayId(stdout: string): number {
  // Desktop 'TurboVNC: t001:2 (2001213077)' started on display t001:2
  // Desktop 'TurboVNC: cn1:21 (demo_admin)' started on display cn1:21
  const regex = /^Desktop '.*' started on display .*:(\d+)$/;

  const lines = stdout.split("\n");

  for (const line of lines) {
    const matches = regex.exec(line);
    if (!matches) {
      continue;
    }

    return +matches[1];
  }

  throw new Error("Error parsing display id");
}

export function displayIdToPort(displayId: number): number {
  return DISPLAY_ID_PORT_DELTA + displayId;
}

export function portToDisplayId(port: number): number {
  return port - DISPLAY_ID_PORT_DELTA;
}

File: apps/scowd/internal/desktop/turbovnc.go
package desktop

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

const DISPLAY_ID_PORT_DELTA = 5900

// ParseOTP takes a string stderr, looks for a line starting with a specific indicator,
// and returns the substring following the indicator as the OTP.
// If no line starts with the indicator, it returns an error.
func ParseOTP(stderr string) (string, error) {
	indicator := "Full control one-time password: "
	lines := strings.Split(stderr, "\n")

	for _, line := range lines {
		if strings.HasPrefix(line, indicator) {
			return strings.TrimSpace(line[len(indicator):]), nil
		}
	}

	return "", errors.New("error parsing OTP")
}

// parseDisplayId takes a string stdout, looks for a line that matches
// a specific regular expression pattern, and extracts the display ID as an integer.
// If no matching line is found, it returns an error.
func ParseDisplayID(stdout string) (int, error) {
	regex := regexp.MustCompile(`^Desktop '.*' started on display .*:(\d+)$`)
	lines := strings.Split(stdout, "\n")

	for _, line := range lines {
		matches := regex.FindStringSubmatch(line)
		if matches != nil {
			// Convert the captured group to an integer
			displayId, err := strconv.Atoi(matches[1])
			if err != nil {
				// Handle the case where the conversion fails
				return 0, fmt.Errorf("error converting display id to integer: %v", err)
			}
			return displayId, nil
		}
	}

	// Return an error if no matching line is found
	return 0, fmt.Errorf("error parsing display id from %s", stdout)
}

func DisplayIDToPort(displayID int) int {
	return DISPLAY_ID_PORT_DELTA + displayID
}
